### PR TITLE
graphics for S60/P & TS101

### PR DESCRIPTION
**Pull request by discip**
@Ralim
Here you go. 😀

Tried to implement all the necessary graphics for **`S60/P`** & **`TS101`**.

I know this will not work until you do all the necessities around it.
At least I hope the way I tried to implement this makes any sense. 😅

Not every graphic is pretty yet, but at least it should be testable.

Please let me know what you think.

menu animations||
-:|-
power|Power
soldering|Soldering
sleep|Sleep
UI|UI
advanced|Advanced

<table>
    <thead>
        <tr>
            <th colspan=4>statics</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td colspan=2></td>
            <td align="center"><b>S60/P</b></td>
            <td align="center"><b>TS101</b></td>
        </tr>
        <tr>
            <td colspan=2>A-button</td>
            <td>A-button</td>
            <td>A-button</td>
        </tr>
        <tr>
            <td colspan=2>B-button</td>
            <td colspan=2 align="center">B-button</td>
        </tr>
        <tr>
            <td colspan=2>missing tip</td>
            <td>missing tip</td>
            <td>missing tip</td>
        </tr>
        <tr>
            <td colspan=2>unavailable</td>
            <td colspan=2 align="center">unavailable</td>
        </tr>
        <tr>
            <td rowspan=2>bootlogo</td>
            <td>once</td>
            <td colspan=2 align="center">once</td>
        </tr>
        <tr>
            <td>loop</td>
            <td colspan=2 align="center">loop</td>
        </tr>
    </tbody>
</table>

### CODE REVIEW

**Comment by discip** [comment]
@Ralim
Don't know if you had a chance to observe the changed pics.
I assume: Most likely not ! 😀
Anyway now there are attached gifs ...

Hope you like it.

**Comment by Ralim** [comment]
Amazing work, suppppppppeeerrrr appreciate it.

Haven't had a chance to test on hardware yet but when I get some downtime I will do some quick patching to get some of this in place asap 🤞

Thank you so much for this ❤️

**Comment by discip** [comment]
@Ralim
What am I doing wrong here, why is it not compiling for the **`S60`** Or the **`TS101`**? 🤷🏻‍♂️
Any idea?

Also https://github.com/Ralim/IronOS/discussions/1806#discussioncomment-7720404. 😁

**Comment by Ralim** [comment]
Unsure, might need to re-base on dev rather than merge. Not quite sure tbqh.

File: source/Core/Drivers/Font.h [hunk 1]
```diff
@@ -8,863 +8,2274 @@
...
-const uint8_t SettingsMenuIcons[][32 * 3] = {
-    // Power
-    // 3 frames
-    {
-        // Power 1st frame
-        // width = 16
+     * 21x32 icons
+     * 53 * 3 = Frame size * Frame count
+     * */
+    const uint8_t SettingsMenuIcons[][53 * 3] = {
```
Comment (ia) [inline source/Core/Drivers/Font.h:183]: TL; DR: replace `53 * 3` by `252`. It will fix builds but I'm not sure if it will work on the hardware.

Longer version: I'm not an expert but a bit of write-up on that, if I may, to help. So, [log says](https://github.com/Ralim/IronOS/actions/runs/7159972697/job/19493827042?pr=1819) there is a compilation error. [Looking a bit above](https://github.com/Ralim/IronOS/actions/runs/7159972697/job/19493827042?pr=1819#step:8:41) there is a piece of additional information:
```
./Core/Drivers/Font.h:1523:5: error: too many initializers for 'const uint8_t [159]' {aka 'const unsigned char [159]'}
```
Basically, highlighted line above:
```
const uint8_t SettingsMenuIcons[][53 * 3]
```
tells that the second dimension of two-dimensional array `SettingsMenuIcons` is going to be stuffed with _at least_ `159` elements. However, the compilation error from log says that _actual_ amount of elements which are going to be stuffed into - _too many_.
Looking through [this version](https://github.com/discip/IronOS/blob/graphics-for-S60/P-%26-TS101/source/Core/Drivers/Font.h) of `Font.h` with the help of bash one liners with sed/grep/awk/etc, amount of elements (max) _roughly_ is `252`. Hence, replacing `53 * 3` by `252` _should_ do the trick. At least I did manage to reproduce the issue locally and then increasing the space for the second array up to `252` did make build successful again. However, I'm not sure about this _woodoo magic numbers_. But it seems that since this part is under `define` for specific models (with bigger OLEDs), I guess the other models shouldn't be affected anyway. :)

And, to be honest, now I'm really curious myself will firmware with this modification work on the real S60/TS101 or there will be _some_ glitches with display.

If some of that somehow correct, then the question now is how to properly arrange that transition from `53 * 3` to `252` from the point of code style.

It seems that this file is [at least partially] generated with some other scripts. If so, unfortunately this is well-known tradeoff for automagicalization: with static code you have to test & debug static code, with automagic code you have to test & debug scripts *AND* the generated code. ;)

I hope some of that will help anyway.

Comment (discip) [reply source/Core/Drivers/Font.h]: Thank you for finding the culprit!
However, I'm not sure if it will work either. 😅

**Review by tsndr — APPROVED** [review]

**Comment by discip** [comment]
@Ralim
> Haven't had a chance to test on hardware yet but when I get some downtime I will do some quick patching to get some of this in place asap 🤞

Have you had time to test this in the mean time?
Is this needed to be merged before #1749, or do you think it would be best to do this afterwards?

Also:
Have you any advice for me about flashing my **`S60P`** with IronOS for **`S60`**?
I'm still somewhat hesitant about doing it, due to the risk of bricking my precious **`S60P`**.

**Comment by discip** [comment]
One more thing about flashing the **`S60P`**:
As long as I will not be able to flash **`IronOS`**, I will not be able to test if the graphics are working. 😅

**Comment by dev0stranger** [comment]
Hello! I have TS101 solder and i am interested in your pull) How i can install your firmware?

**Comment by discip** [comment]
@dev0stranger
This PR needs to be tested first.
I don't own a **`TS101`** so I couldn't do this.

**Comment by brushmate** [comment]
I somehow managed to compile your branch for my TS101 by rebasing your branch on the dev branch, but the icons are not being displayed correctly. I guess this is because support for bigger displays has not been added yet?

IMG_20240211_210201

**Comment by discip** [comment]
@Jonny-T-cyber
https://forum.minidso.com/forum.php?mod=viewthread&tid=4937&extra=page%3D1

**Comment by discip** [comment]
> I somehow managed to compile your branch for my TS101 by rebasing your branch on the dev branch, but the icons are not being displayed correctly. I guess this is because support for bigger displays has not been added yet?

As you may have noticed in your photo only a portion of the OLED but the individual graphics are intended for the whole screen.
So we have to wait until the support for the entire OLED gets implemented.

**Comment by discip** [comment]
@brushmate
> I somehow managed to compile your branch for my TS101 by rebasing your branch on the dev branch, but the icons are not being displayed correctly. I guess this is because support for bigger displays has not been added yet?

As you may have noticed in your photo only a portion of the OLED but the individual graphics are intended for the whole screen.
So we have to wait until the support for the entire OLED gets implemented.

**Comment by ArturoGuerra** [comment]
Besides the graphics what else is needed? @Ralim

**Comment by Ralim** [comment]
This will be pulled in via the branch [for splitting out rendering for different screen sizes](https://github.com/Ralim/IronOS/tree/gui-multi-screen). I'm slowly working on it as life allows me to.

**Comment by ArturoGuerra** [comment]
> This will be pulled in via the branch [for splitting out rendering for different screen sizes](https://github.com/Ralim/IronOS/tree/gui-multi-screen). I'm slowly working on it as life allows me to.

okay I'll have a look in a bit and see if I can help with that, also for the standardized power interface do you want to include all currently supported usb protocols so QC and USB-PD or just USB-PD?

**Comment by Ralim** [comment]
Once I finish splitting all the render/* functions out (in a day or two) I'll start a draft PR to track it.
When that happens then  basically need to update every single render function to layout on the larger screen correctly 😓 

I only intend to implement PD for all devices going forward if I get any choice in it. QC is a dead standard and its a shit-show for compatibility as well. Much prefer only focusing on standards that are open and _properly_ standardised.

**Comment by discip** [comment]
@Ralim
Any news so far? 😊

**Comment by Ralim** [comment]
> Any news so far? 😊

I've now pulled these into #1883, and will slowly be working over there to update all the drawing commands to draw the right size / location

**Comment by discip** [comment]
@Ralim
Hey Ben,
thank you for keeping this going. 😃👍

Could you please help finding the cause, why this doesn't compile for the **`S60P`**, but for the **`TS101`** it does?

thanks in advance

**Comment by ia** [comment]
> Could you please help finding the cause, why this doesn't compile for the **`S60P`**, but for the **`TS101`** it does?

Hello.

_It seems_ [there is no declaration](https://github.com/Ralim/IronOS/actions/runs/8536576398/job/23385375696#step:8:24) of `ButtonA` and related stuff in `Font.h` for `S60P` model explicitly. What's going on in `Font.h` is pure magic for me (I mean those constants), but apparently between all of these `#ifdef MODEL / #else / #endif` there is a missing conditional branch for macro pre-processor managing values of `ButtonA` and `disconnectedTip`.

Pre-processor goes [here](https://github.com/discip/IronOS/blob/7dae223ad54bed3b1296c0f4d1c9f414c571ec4c/source/Core/Drivers/Font.h#L97) for `S60P` first, but then there is conditions like:
[`#ifdef MODEL_S60`](https://github.com/discip/IronOS/blob/7dae223ad54bed3b1296c0f4d1c9f414c571ec4c/source/Core/Drivers/Font.h#L98C5-L98C20)  
and  
[`#elif defined MODEL_TS101`](https://github.com/discip/IronOS/blob/7dae223ad54bed3b1296c0f4d1c9f414c571ec4c/source/Core/Drivers/Font.h#L123)  
but there is no values of the variables for `S60P`, because there is no such condition for `S60P` to define related values.

Basically, since support for `S60P` has been implemented, now `Font.h` should be updated and ported to support for `S60P`.

_**Probably**_ for fix you should make [this line](https://github.com/discip/IronOS/blob/7dae223ad54bed3b1296c0f4d1c9f414c571ec4c/source/Core/Drivers/Font.h#L98C1-L98C21) to look something like this:
`#if defined(MODEL_S60) || defined(MODEL_S60P)`
but I'm not 100% sure that the values from `S60` variables will be valid for `S60P` model.

Hope it will help.

**Comment by discip** [comment]
Thank you very much for the effort you put into describing the issue so precisely!

You are absolutely right. I forgot to add **`S60P`** to Font.h as well. 🤦‍♂️

**Comment by ia** [comment]
> Thank you very much for the effort you put into describing the issue so precisely!
> 
> > #if defined(MODEL_S60) || defined(MODEL_S60P)
> 
> You are absolutely right. That's the line I missed. 🤦‍♂️

I just was happy to help with everything I can, as always. Glad it worked. :)

**Comment by discip** [comment]
This was implemented in #1888.
